src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -264,9 +264,29 @@ def find_candidates(
             if ireq is not None:
                 ireqs.append(ireq)
 
+        for link in constraint.links:
+            # This part only has to convert an ireq from the requirements to a link.
+            # If there's only links, either it satisfies, or it doesn't.
+            if not ireqs:
+                break
+            ireq = ireqs[0]
+            if link.is_wheel:
+                wheel = Wheel(link.filename)
+                if not wheel.supported(self._finder.target_python.get_tags()):
+                    continue
+            candidate = self._make_candidate_from_link(
+                link,
+                extras=frozenset(),
+                template=ireq,
+                name=canonicalize_name(ireq.name) if ireq.name else None,
+                version=None,
+            )
+            if candidate is not None:
+                explicit_candidates.add(candidate)
```

Comment:
A `None` return value means the candidate fails to build, and we’re screwed if that happens (the constraint forces using that URL, nothing would work if it fails). So maybe we can just give up here instead (`return ()`)?